File: files_processors/poalim_processor.py
from files_processors.raw_files import HeaderMapItem
from files_processors.raw_files import RawCSVFile
from files_processors.raw_files import YnabCsvFields

INFLOW_KEY = "זכות"
OUTFLOW_KEY = "חובה"
POALIM = "shekel"

class PoalimProcessor(RawCSVFile):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._header_mapping = [
            HeaderMapItem(source="\ufeffתאריך", target=YnabCsvFields.DATE.value),
            HeaderMapItem(source="תיאור הפעולה", target=YnabCsvFields.PAYEE.value),
            HeaderMapItem(source="אסמכתא", target=YnabCsvFields.MEMO.value),
            HeaderMapItem(source="סכום", target=YnabCsvFields.OUTFLOW.value),
            HeaderMapItem(source="זכות", target=YnabCsvFields.INFLOW.value),
        ]

        self._json_mapping = {
            YnabCsvFields.DATE_KEY.value: "\ufeffתאריך",
            YnabCsvFields.PAYEE_KEY.value: "תיאור הפעולה",
            YnabCsvFields.MEMO_KEY.value: "אסמכתא",
            YnabCsvFields.AMOUNT_KEY.value: "סכום",
        }

        self._balance_field = "יתרה לאחר פעולה"
        self._body_rows = self._get_body_rows()

    def get_balance(self):
        return float(self._body_rows[-1]["balance"])

    @staticmethod
    def identify_account(file, accounts=[]):
        if POALIM in file.name:
            return "poalim"
        return None

    def _get_body_rows(self):
        return [self._getRowObject(row) for row in self._body_rows]

    def _getRowObject(self, row):
        year, month, day = self._split_date(row[self._json_mapping[YnabCsvFields.DATE_KEY.value]])
        outflow = row[OUTFLOW_KEY]
        inflow = row[INFLOW_KEY]
        return {
            self._header_mapping[0].source: self._get_ynab_date(year, month, day),
            self._header_mapping[1].source: row[self._json_mapping[YnabCsvFields.PAYEE_KEY.value]],
            self._header_mapping[2].source: row[self._json_mapping[YnabCsvFields.MEMO_KEY.value]],
            self._header_mapping[3].source: self._getYnabAmount((float(outflow or inflow)),
                                                                  outflow=inflow == ""),
            "balance": row[self._balance_field]
        }

    def _split_date(self, value):
        return value.split("-")

File: app/app/utils/filesProcessors/poalimProcessor.ts
import RawCSVFile from "~/utils/filesProcessors/rawFiles/RawCSVFile";
import {YnabCsvFields} from "~/utils/filesProcessors/rawFiles/RawFile";

const INFLOW_KEY = "זכות";
const OUTFLOW_KEY = "חובה";
const POALIM = "shekel";

export class PoalimProcessor extends RawCSVFile {
    constructor(kwargs: any) {
        super(kwargs);
        this._header_mapping = [
            new HeaderMapItem({source: "\ufeffתאריך", target: YnabCsvFields.DATE}),
            new HeaderMapItem({source: "תיאור הפעולה", target: YnabCsvFields.PAYEE}),
            new HeaderMapItem({source: "אסמכתא", target: YnabCsvFields.MEMO}),
            new HeaderMapItem({source: "סכום", target: YnabCsvFields.OUTFLOW}),
            new HeaderMapItem({source: "זכות", target: YnabCsvFields.INFLOW}),
        ];

        this._json_mapping = {
            [YnabCsvFields.DATE_KEY]: "\ufeffתאריך",
            [YnabCsvFields.PAYEE_KEY]: "תיאור הפעולה",
            [YnabCsvFields.MEMO_KEY]: "אסמכתא",
            [YnabCsvFields.AMOUNT_KEY]: "סכום",
        };

        this._balance_field = "יתרה לאחר פעולה";
        this._body_rows = this._get_body_rows();
    }

    getBalance() {
        return parseFloat(this._body_rows[this._body_rows.length - 1]["balance"]);
    }

    static identifyAccount(file: File, accounts = []) {
        if (file.name.includes(POALIM)) {
            return "poalim";
        }
        return null;
    }

    _getBodyRows() {
        return this._body_rows.map((row: any) => this._getRowObject(row));
    }

    _getRowObject(row: any) {
        const [year, month, day] = row[this._json_mapping[YnabCsvFields.DATE_KEY]].split("-");
        const outflow = row[OUTFLOW_KEY];
        const inflow = row[INFLOW_KEY];
        return {
            [this._header_mapping[0].source]: this._get_ynab_date(year, month, day),
            [this._header_mapping[1].source]: row[this._json_mapping[YnabCsvFields.PAYEE_KEY]],
            [this._header_mapping[2].source]: row[this._json_mapping[YnabCsvFields.MEMO_KEY]],
            [this._header_mapping[3].source]: this._getYnabAmount(parseFloat(outflow || inflow),
                outflow === ""),
            "balance": row[this._balance_field]
        };
    }
}

module.exports = PoalimProcessor;
